File: 249. Group Shifted Strings/groupStrings.js
/**
 * @param {string[]} strings
 * @return {string[][]}
 */
const groupStrings = (strings) => {
  const group = {};
  const shiftLetter = (letter, shift) => {
    const chCode = (letter.charCodeAt(0) - shift + 26) % 26 + 'a'.charCodeAt(0);
    return chCode;
  }

  const getHash = (str) => {
    const shift = str.charCodeAt(0);
    const strCodes = [...str].map(it => shiftLetter(it, shift));
    return String.fromCharCode(...strCodes);
  }

  for (let i = 0; i < strings.length; i++) {
    const key = getHash(strings[i]);
    if (!group[key]) {
      group[key] = [];
    }
    group[key].push(strings[i])
  }

  return Object.values(group)
};


const strings = ["abc", "bcd", "acef", "xyz", "az", "ba", "a", "z"];
console.log('groupStrings: ', groupStrings(strings));

File: 249. Group Shifted Strings/groupStrings.py
from typing import List

class Solution:
  def groupStrings(self, strings: List[str]) -> List[List[str]]:
    group = {}
    def shift_letter(letter , shift):
      c = (ord(letter) - shift) % 26 + ord('a')
      return chr(c)
    
    def get_hash(string):
      shift = ord(string[0])
      letters_arr = [shift_letter(letter, shift) for letter in string]
      return ''.join(letters_arr)
    
    for s in strings:
      key = get_hash(s)
      if key not in group:
        group[key] = []
      group[key].append(s)

    return list(group.values())

 
strings = ["abc", "bcd", "acef", "xyz", "az", "ba", "a", "z"]
solution = Solution()
print('groupStrings: ', solution.groupStrings(strings))
